docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+dialog a little more future proof.
+
+To drive home the synchronous nature of the `window.prompt` notice that the time
+stops ticking when the prompt dialog is displayed. Lets fix that by making our
+own prompt. Since our dialog is just DOM manipulation the page won't be blocked
+while waiting for user input.
+
+First add the prompt dialog to the HTML:
+
+```html
+<style type="text/css">
+  #dialog {
+    width:      200px;
+    margin:     auto;
+    padding:    10px;
+    border:     thin solid black;
+    background: lightgreen;
+  }
+  .hidden {
+    display: none;
+  }
+</style>
+<div id="dialog" class="hidden">
+  <div class="message">foobar</div>
+  <input type="text">
+  <div>
+    <button class="ok">Ok</button>
+    <button class="cancel">Cancel</button>
+  </div>
+</div>
+```
+
+We will want to keep the same API so our change will be only to the
+`promisePrompt`. It will find the dialog DOM elements, attach events to the
+elements, show the dialog box, return a promise that is resolved based on the
+attached events, and finally detaches the events and cleans up after itself
+(hiding the dialog box for another use later).
+
+```javascript
+function promptPromise(message) {
+  var dialog       = document.getElementById('dialog');
+  var input        = dialog.querySelector('input');
+  var okButton     = dialog.querySelector('button.ok');
+  var cancelButton = dialog.querySelector('button.cancel');
+  var resolver, rejector;
+
+  dialog.querySelector('.message').innerHTML = '' + message;
+  dialog.className = '';
+
+  function handleOk() {
+    resolver(input.value);
+  }
+
+  function handleCancel() {
+    rejector();
+  }
+
+  okButton.addEventListener('click', handleOk);
+  cancelButton.addEventListener('click', handleCancel);
+
+  return new Promise(function(resolve, reject) {
+    resolver = resolve;
+    rejector = reject;
+  })
+  .finally(function() {
+    okButton.removeEventListener('click', handleOk);
+    cancelButton.removeEventListener('click', handleCancel);
+    dialog.className = 'hidden';
+  });
+}
+```
+
+[Run example on JSBin](http://jsbin.com/fucofu/edit?js,output)
+
+Now when the user presses the **Set Name** button the clock continues to update
+while the dialog is visible.
+
+Because the `removeEventListener` requires a reference to the original function
+that was used with the `addEventListener` it makes it difficult to clean up
+after itself without storing the references in a scope higher then the promise
+resolution function. This pattern of storing the `resolve` and `reject`
+callbacks outside the scope of the promise function is called [the **deferred**
+anti-pattern](api/anti-patterns.html#the-deferred-anti-pattern). It is an
+*anti-pattern* because it is often used when it is not needed. However, in this
+case it is the only way to accomplish the clean-up requirements imposed by the
+DOM API (namely the need for the original function reference to
+`removeEventListener`).
```

Comment:
That's not really true. The deferred anti pattern is about explicit construction (not extracting them outside), you can just put them in named functions inside the promise construction. See - stackoverflow.com/questions/23803743/what-is-the-explicit-promise-construction-antipattern-and-how-do-i-avoid-it 

This is just promisifying an API (you needed the promise constructor) which you need to do because it's not in a standard format.